### DISCUSSION TITLE
Regarding the current drive cycle.

### DISCUSSION THREAD

**Discussion by manjunathnilugal** [pybamm-team/PyBaMM#2746] [Q&A] [answered]
### Discussed in https://github.com/pybamm-team/PyBaMM/discussions/2742

<div type='discussions-op-text'>

<sup>Originally posted by **manjunathnilugal** March  5, 2023</sup>
Hello everyone,
I am trying to run a current drive cycle, at some current values the voltage reaches maximum cut-off voltage 4.15 V. In that case I want to change the current value to "0 A" and maintain constant voltage until next discharge current value.
I am attaching the code which I am trying to execute.

```
import pybamm
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

model = pybamm.lithium_ion.SPM()
V = model.variables["Terminal voltage [V]"]

dt = 3600
time = 0
end_time = 11*3600

geometry = model.default_geometry
param = pybamm.ParameterValues("Chen2020")
#model.events.append(pybamm.Event('V', V - 4.1))

Vmin = 2.5
Vmax = 4.15

param['Lower voltage cut-off [V]'] = Vmin
param['Upper voltage cut-off [V]'] = Vmax

param['Current function [A]'] = "[input]"

param.process_model(model)
param.process_geometry(geometry)

# set mesh
mesh = pybamm.Mesh(geometry, model.default_submesh_types, model.default_var_pts)

# discretise model
disc = pybamm.Discretisation(mesh, model.default_spatial_methods)
disc.process_model(model)

I = np.array([2, 1, -2, -1, -3, -1, -2, 3, 2, 5, 4])
    
step_solver = pybamm.CasadiSolver()
step_solution = None

for i,I in enumerate(I):
    Vt = np.array([0])
    
    if  Vt[-1] >= 4.2:
        inputs = {'Current function [A]': 0.001}
    else:
        inputs = {'Current function [A]': I}
    
    step_solution = step_solver.step(step_solution, model, dt=dt, inputs=inputs, npts=100)
    V = step_solution["Terminal voltage [V]"].entries
    Vt = np.concatenate((Vt,V))
    It = step_solution["Current [A]"].entries

step_solution.plot()
```
I am not able to iterate the current values in the step solution, the simulation exits after the cut-off voltage is reached. Is there any method to keep the track of the voltage variation in middle of simulation ?
<img width="134" alt="image" src="https://user-images.githubusercontent.com/82285213/222927900-63be6338-123e-49ba-8091-a49900841165.png">

Instead of creating a interpolant function for current, I want to pass current individually after checking the final voltage value after each iteration.

```
if final voltage value is >= 4.2V:
      Current is 0.001 A
else:
      Current is [i] A
```
Is it possible to do using pybamm.</div>

**Comment by brosaplanella** [comment, marked as answer]
I would suggest using experiments for that (see [this example](https://github.com/pybamm-team/PyBaMM/blob/develop/examples/notebooks/Getting%20Started/Tutorial%205%20-%20Run%20experiments.ipynb)). Note, however, that you can't set both the current and voltage. That is, if you set the voltage to 4.15 V, the current will be whatever value that ensures that voltage.

**Reply by Mrzhang-hub** [reply]
Due to some reason, I have to input the current though input function. And this process lasts about 1 second. If the current is too large, the terminal voltage will excess the maximum voltage and it will give an error. Is there a way to set the voltage to its maximum voltage and output this value without giving an error?

**Reply by manjunathnilugal** [reply]
You can refer to, 
[https://github.com/pybamm-team/PyBaMM/discussions/2742]
You might get some idea, which might help you.

**Reply by Mrzhang-hub** [reply]
Thanks for your help and problem has been solved according to this post. But the solving speed is too slow via `pybamm.experiment()` compared to `sim.step()`. Have you met the same problem?

**Reply by manjunathnilugal** [reply]
yes, problem solving is very slow (time and ram consuming)
I have raised another discussion, in order to reduce the computation timing.
[https://github.com/pybamm-team/PyBaMM/discussions/2687],

**Reply by Mrzhang-hub** [reply]
Amazing, I will keep track of that. Thanks again!

**Reply by manjunathnilugal** [reply]
@Mrzhang-hub if you find better way to simulate faster, please post here.

**Reply by Mrzhang-hub** [reply]
I get the same error and could not solve this for now.  
`AttributeError: 'Solution' object has no attribute 'steps'`
